### PR TITLE
mgmt Storage, support `allowCrossTenantReplication` and `defaultToOAuthAuthentication` for `StorageAccount`

### PR DESCRIPTION
**Pull request by v-hongli1**
relate to [issue#37451](https://github.com/Azure/azure-sdk-for-java/issues/37451)

# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java:246]: Any doc or evidence to support this behavior(if null, then true)?

**Comment by XiaofeiCao** [comment]
Thanks, please also add in changelog.md `Feature Request` section about the changes.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:951]: the next stage of storage account update

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java
Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java:246]: Yeah, please check Portal or CLI.

If they didn't show, it be safer just use `Boolean` and allow null.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:572]: We usually do `enableCrossTenantReplication()`, if default be false.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:583]: Same, maybe `enableDefaultToOAuthAuthentication()`.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java]: Now fixed to `return ResourceManagerUtils.toPrimitiveBoolean(this.innerModel().allowCrossTenantReplication());`.  The another one also had the same fix.

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java]: Could you confirm whether default (when send `null`), it is true or false?

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:583]: That means you need to use the "enable" and "disable" methods to enable and disable "allowCrossTenantReplication" and "defaultToOAuthAuthentication" respectively, is that right?

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:583]: Yes. Depends on the meaning of `null`.

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:583]: Yes, if default is false, then in create phase, we only need `enableCrossTenantReplication()`(or `enableDefaultToOAuthAuthentication()`).

In update phase, we probably need both `enableCrossTenantReplication()` and `disableCrossTenantReplication()`.

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (XiaofeiCao) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:580]: Would you let us know whether the default(if `null` is specified) is true or false?

If true, then we probably only need `disableAllowCrossTenantReplication()` in create phase. Otherwise, we only need `enableAllowCrossTenantReplication()` in create phase. (update mode need both no matter what)

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: I rechecked the status of the `Storage Account` in the `Azure Portal`. The latest version is now consistent with the initial settings and changes of status in `Azure Portal`.

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: The default status of `AllowCrossTenantReplication` is true, so only keep `disableAllowCrossTenantReplication()` in create.
The default status of `defaultToOAuthAuthentication` is false, so only keep `enableDefaultToOAuthAuthentication()` in create.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: nit, not sure `disableAllowCrossTenantReplication` or `disableCrossTenantReplication`.

**Review by weidongxu-microsoft — APPROVED** [review]
LGTM.
Please wait for 2 approval here (as it is handwritten code).

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: From the meaning of `allowCrossTenantReplication`, this is an allow setting, so I think the `disableAllowCrossTenantReplication` is more appropriate.

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: Agree, `disableCrossTenantReplication` is better. `disable`(`enable`) and `allow` seem duplicate.

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: `enableAllowCrossTenantReplication` rename to `enableCrossTenantReplication`
`disableAllowCrossTenantReplication` rename to `disableCrossTenantReplication`
Is that right ?

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: Let us check more doc. @v-hongli1 If you have doc from learn.microsoft.com, would you paste the link here?

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: "enableAllow*" has the problem of having 2 verb.

But if we remove the "Allow" it may look like we always do cross tenant replication, while this may not be the exact case.

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: Or perhaps, `allowCrossTenantReplication` and `disallowCrossTenantReplication`.

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: Yeah, that looks good.

Let's wait a bit for Hong's doc link,

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: I haven't found any documentation yet, but I have found similar code examples. I don't know if they can be used as a reference.
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/implementation/StorageAccountImpl.java#L529

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:573]: No. Not Java code.

I mean service doc about the feature e.g. https://learn.microsoft.com/en-us/azure/storage/blobs/object-replication-prevent-cross-tenant-policies?tabs=azure-cli

Let's use Xiaofei's suggestion above. https://github.com/Azure/azure-sdk-for-java/pull/37460#discussion_r1378359912

Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java]: `enableAllowCrossTenantReplication` rename to `allowCrossTenantReplication` and`disableAllowCrossTenantReplication` rename to `disallowCrossTenantReplication` in the new version.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:945]: nit, `.` at the end of the sentence. Please double check in your javadoc.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:964]: same here

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:567]: same here.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java:964]: Fixed in the new version.

File: sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java
Comment (v-hongli1) [reply sdk/resourcemanager/azure-resourcemanager-storage/src/main/java/com/azure/resourcemanager/storage/models/StorageAccount.java]: Fixed in the new version.

**Review by XiaofeiCao — APPROVED** [review]

**Pull request merged**
